File: frontend/src/types/sandbox.ts
export type SandboxStatus = 'active' | 'paused' | 'stopped';
export type DeploymentStatus = 'pending' | 'running' | 'paused' | 'completed' | 'failed';
export type TransactionType = 'deposit' | 'withdraw' | 'buy' | 'sell' | 'dividend' | 'fee';

export const SANDBOX_STATUS_LABELS: Record<SandboxStatus, string> = {
  active: '运行中',
  paused: '已暂停',
  stopped: '已停止',
};

export const DEPLOYMENT_STATUS_LABELS: Record<DeploymentStatus, string> = {
  pending: '待运行',
  running: '运行中',
  paused: '已暂停',
  completed: '已完成',
  failed: '失败',
};

export const TRANSACTION_TYPE_LABELS: Record<TransactionType, string> = {
  deposit: '入金',
  withdraw: '出金',
  buy: '买入',
  sell: '卖出',
  dividend: '分红',
  fee: '手续费',
};

export interface SandboxAccount {
  id: number;
  user_id: number;
  name: string;
  description?: string;
  initial_capital: number;
  current_cash: number;
  total_value: number;
  status: SandboxStatus;
  created_at: string;
  updated_at: string;
}

export interface SandboxAccountCreate {
  name: string;
  description?: string;
  initial_capital?: number;
}

export interface SandboxAccountUpdate {
  name?: string;
  description?: string;
  status?: SandboxStatus;
}

export interface SandboxAccountListParams {
  page?: number;
  size?: number;
}

export interface SandboxAccountListResponse {
  items: SandboxAccount[];
  total: number;
  page: number;
  size: number;
}

export interface SandboxPosition {
  id: number;
  account_id: number;
  stock_code: string;
  stock_name?: string;
  quantity: number;
  avg_cost: number;
  current_price: number;
  market_value: number;
  unrealized_pnl: number;
  realized_pnl: number;
  updated_at: string;
}

export interface SandboxTransaction {
  id: number;
  account_id: number;
  deployment_id?: number;
  transaction_type: TransactionType;
  stock_code?: string;
  stock_name?: string;
  quantity?: number;
  price?: number;
  amount: number;
  commission: number;
  description?: string;
  created_at: string;
}

export interface SandboxDeployment {
  id: number;
  account_id: number;
  strategy_id: number;
  name: string;
  status: DeploymentStatus;
  start_date: string;
  end_date?: string;
  stock_pool?: string[];
  parameters?: Record<string, unknown>;
  allocation_ratio: number;
  last_run_date?: string;
  last_run_result?: Record<string, unknown>;
  error_message?: string;
  created_at: string;
  updated_at: string;
}

export interface SandboxDeploymentCreate {
  strategy_id: number;
  name: string;
  start_date: string;
  end_date?: string;
  stock_pool?: string[];
  parameters?: Record<string, unknown>;
  allocation_ratio?: number;
}

export interface SandboxDailyValue {
  id: number;
  account_id: number;
  date: string;
  total_value: number;
  cash: number;
  position_value: number;
  daily_return?: number;
  cumulative_return?: number;
  benchmark_return?: number;
}

export interface SandboxAccountDetail {
  account: SandboxAccount;
  positions: SandboxPosition[];
  recent_transactions: SandboxTransaction[];
  deployments: SandboxDeployment[];
  daily_values: SandboxDailyValue[];
}

export interface DepositRequest {
  amount: number;
  description?: string;
}

export interface ResetAccountRequest {
  initial_capital?: number;
}

File: backend/app/models/sandbox.py
from datetime import datetime
from enum import Enum
from typing import Optional

from sqlalchemy import (
    JSON,
    Boolean,
    Column,
    DateTime,
    Enum as SQLEnum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    Date,
    UniqueConstraint,
)
from sqlalchemy.orm import relationship

from app.models.base import Base


class SandboxStatus(str, Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    STOPPED = "stopped"


class DeploymentStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"


class TransactionType(str, Enum):
    DEPOSIT = "deposit"
    WITHDRAW = "withdraw"
    BUY = "buy"
    SELL = "sell"
    DIVIDEND = "dividend"
    FEE = "fee"


class SandboxAccount(Base):
    __tablename__ = "sandbox_accounts"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    name = Column(String(100), nullable=False)
    description = Column(Text, nullable=True)
    initial_capital = Column(Float, default=1000000.0)
    current_cash = Column(Float, default=1000000.0)
    total_value = Column(Float, default=1000000.0)
    status = Column(SQLEnum(SandboxStatus), default=SandboxStatus.ACTIVE)
    version = Column(Integer, default=0, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    positions = relationship("SandboxPosition", back_populates="account", cascade="all, delete-orphan")
    transactions = relationship("SandboxTransaction", back_populates="account", cascade="all, delete-orphan")
    deployments = relationship("SandboxDeployment", back_populates="account", cascade="all, delete-orphan")
    daily_values = relationship("SandboxDailyValue", back_populates="account", cascade="all, delete-orphan")


class SandboxPosition(Base):
    __tablename__ = "sandbox_positions"
    __table_args__ = (
        UniqueConstraint("account_id", "stock_code", name="uq_sandbox_position_account_stock"),
    )

    id = Column(Integer, primary_key=True, index=True)
    account_id = Column(Integer, ForeignKey("sandbox_accounts.id"), nullable=False)
    stock_code = Column(String(20), nullable=False)
    stock_name = Column(String(50), nullable=True)
    quantity = Column(Integer, default=0)
    avg_cost = Column(Float, default=0.0)
    current_price = Column(Float, default=0.0)
    market_value = Column(Float, default=0.0)
    unrealized_pnl = Column(Float, default=0.0)
    realized_pnl = Column(Float, default=0.0)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    account = relationship("SandboxAccount", back_populates="positions")


class SandboxTransaction(Base):
    __tablename__ = "sandbox_transactions"

    id = Column(Integer, primary_key=True, index=True)
    account_id = Column(Integer, ForeignKey("sandbox_accounts.id"), nullable=False)
    deployment_id = Column(Integer, ForeignKey("sandbox_deployments.id"), nullable=True)
    transaction_type = Column(SQLEnum(TransactionType), nullable=False)
    stock_code = Column(String(20), nullable=True)
    stock_name = Column(String(50), nullable=True)
    quantity = Column(Integer, nullable=True)
    price = Column(Float, nullable=True)
    amount = Column(Float, nullable=False)
    commission = Column(Float, default=0.0)
    description = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    account = relationship("SandboxAccount", back_populates="transactions")
    deployment = relationship("SandboxDeployment", back_populates="transactions")


class SandboxDeployment(Base):
    __tablename__ = "sandbox_deployments"

    id = Column(Integer, primary_key=True, index=True)
    account_id = Column(Integer, ForeignKey("sandbox_accounts.id"), nullable=False)
    strategy_id = Column(Integer, ForeignKey("strategies.id"), nullable=False)
    name = Column(String(100), nullable=False)
    status = Column(SQLEnum(DeploymentStatus), default=DeploymentStatus.PENDING)
    start_date = Column(Date, nullable=False)
    end_date = Column(Date, nullable=True)
    stock_pool = Column(JSON, nullable=True)
    parameters = Column(JSON, nullable=True)
    allocation_ratio = Column(Float, default=1.0)
    last_run_date = Column(Date, nullable=True)
    last_run_result = Column(JSON, nullable=True)
    error_message = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    account = relationship("SandboxAccount", back_populates="deployments")
    strategy = relationship("Strategy")
    transactions = relationship("SandboxTransaction", back_populates="deployment")


class SandboxDailyValue(Base):
    __tablename__ = "sandbox_daily_values"
    __table_args__ = (
        UniqueConstraint("account_id", "date", name="uq_sandbox_daily_value_account_date"),
    )

    id = Column(Integer, primary_key=True, index=True)
    account_id = Column(Integer, ForeignKey("sandbox_accounts.id"), nullable=False)
    date = Column(Date, nullable=False)
    total_value = Column(Float, nullable=False)
    cash = Column(Float, nullable=False)
    position_value = Column(Float, nullable=False)
    daily_return = Column(Float, nullable=True)
    cumulative_return = Column(Float, nullable=True)
    benchmark_return = Column(Float, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    account = relationship("SandboxAccount", back_populates="daily_values")
